src/cunumeric/sort/sort.cu

```diff
@@ -0,0 +1,691 @@
+    VAL init_value          = std::numeric_limits<VAL>::max();
+    extract_split_positions<<<num_blocks, THREADS_PER_BLOCK, 0, stream>>>(
+      local_sorted.values.ptr(0),
+      volume,
+      samples.ptr(0),
+      num_usable_samples,
+      split_positions.ptr(0),
+      num_splitters,
+      my_rank);
+  }
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // collect sizes2send, send to rank i: local_sort_data from positions  split_positions[i-1],
+  // split_positions[i] - 1
+  auto size_send = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  {
+    size_t last_position = 0;
+    for (size_t rank = 0; rank < num_ranks - 1; ++rank) {
+      size_t cur_position = split_positions[rank];
+      size_send[rank]     = cur_position - last_position;
+      last_position       = cur_position;
+    }
+    size_send[num_ranks - 1] = volume - last_position;
+  }
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // all2all exchange send/receive sizes
+  auto size_recv = create_buffer<size_t>(num_ranks, Memory::Z_COPY_MEM);
+  CHECK_NCCL(ncclGroupStart());
+  for (int r = 0; r < num_ranks; r++) {
+    CHECK_NCCL(ncclSend(size_send.ptr(r), 1, ncclUint64, r, *comm, stream));
+    CHECK_NCCL(ncclRecv(size_recv.ptr(r), 1, ncclUint64, r, *comm, stream));
+  }
+  CHECK_NCCL(ncclGroupEnd());
+
+  // need to sync as we share values in between host/device
+  CHECK_CUDA(cudaStreamSynchronize(stream));
+
+  // allocate merge targets, data transfer...
+  std::vector<SortPiece<VAL>> merge_buffers(num_ranks);
+
+  for (int i = 0; i < merge_buffers.size(); ++i) {
+    // align buffer to allow data transfer of 16byte blocks
+    auto recv_size_aligned   = get_aligned_size(size_recv[i] * sizeof(VAL));
+    auto buf_size            = (recv_size_aligned + sizeof(VAL) - 1) / sizeof(VAL);
+    merge_buffers[i].values  = create_buffer<VAL>(buf_size, Memory::GPU_FB_MEM);
+    merge_buffers[i].indices = create_buffer<int64_t>(argsort ? buf_size : 0, Memory::GPU_FB_MEM);
+    merge_buffers[i].size    = size_recv[i];
+  }
+  size_t send_pos = 0;
+  CHECK_NCCL(ncclGroupStart());
+  for (int r = 0; r < num_ranks; r++) {
+    CHECK_NCCL(ncclSend(local_sorted.values.ptr(send_pos),
+                        get_aligned_size(size_send[r] * sizeof(VAL)),
+                        ncclInt8,
+                        r,
+                        *comm,
+                        stream));
+    CHECK_NCCL(ncclRecv(merge_buffers[r].values.ptr(0),
+                        get_aligned_size(size_recv[r] * sizeof(VAL)),
+                        ncclInt8,
+                        r,
+                        *comm,
+                        stream));
+    if (argsort) {
+      CHECK_NCCL(
+        ncclSend(local_sorted.indices.ptr(send_pos), size_send[r], ncclInt64, r, *comm, stream));
```

Comment:
this line has the same problem as the above.

Reply:
As far as I understand there is no buffer overrun here - we send exactly the amount of data that we have.

Reply:
Ah I didn't notice that the transfers were aligned by the type. This looks safe now.